core/arch/arm/kernel/secure_partition.c

```diff
@@ -0,0 +1,751 @@
+			 stmm_image, stmm_image_size);
+
+	res = vm_set_prot(&spc->uctx, image_addr, uncompressed_size_roundup,
+			  TEE_MATTR_URX | TEE_MATTR_PR);
+	if (res)
+		return res;
+
+	res = vm_set_prot(&spc->uctx, heap_addr, stmm_heap_size,
+			  TEE_MATTR_URW | TEE_MATTR_PRW);
+	if (res)
+		return res;
+
+	res = vm_set_prot(&spc->uctx, stack_addr, stmm_stack_size,
+			  TEE_MATTR_URW | TEE_MATTR_PRW);
+	if (res)
+		return res;
+
+	res = vm_set_prot(&spc->uctx, sec_buf_addr, stmm_sec_buf_size,
+			  TEE_MATTR_URW | TEE_MATTR_PRW);
+	if (res)
+		return res;
+
+	DMSG("stmm load address %#"PRIxVA, image_addr);
+
+	boot_info = (struct secure_partition_boot_info *)sec_buf_addr;
+	mp_info = (struct secure_partition_mp_info *)(boot_info + 1);
+	*boot_info = (struct secure_partition_boot_info){
+		.h.type = SP_PARAM_SP_IMAGE_BOOT_INFO,
+		.h.version = SP_PARAM_VERSION_1,
+		.h.size = sizeof(struct secure_partition_boot_info),
+		.h.attr = 0,
+		.sp_mem_base = sp_addr,
+		.sp_mem_limit = sp_addr + sp_size,
+		.sp_image_base = image_addr,
+		.sp_stack_base = stack_addr,
+		.sp_heap_base = heap_addr,
+		.sp_ns_comm_buf_base = comm_buf_addr,
+		.sp_shared_buf_base = sec_buf_addr,
+		.sp_image_size = stmm_image_size,
+		.sp_pcpu_stack_size = stmm_stack_size,
+		.sp_heap_size = stmm_heap_size,
+		.sp_ns_comm_buf_size = stmm_ns_comm_buf_size,
+		.sp_shared_buf_size = stmm_sec_buf_size,
+		.num_sp_mem_regions = 6,
+		.num_cpus = 1,
+		.mp_info = mp_info,
+	};
+	mp_info->mpidr = read_mpidr_el1();
+	mp_info->linear_id = 0;
+	mp_info->flags = MP_INFO_FLAG_PRIMARY_CPU;
+	spc->ns_comm_buf_addr = comm_buf_addr;
+	spc->ns_comm_buf_size = stmm_ns_comm_buf_size;
+
+	init_stmm_regs(spc, sec_buf_addr,
+		       (vaddr_t)(mp_info + 1) - sec_buf_addr,
+		       stack_addr + stmm_stack_size, image_addr);
+
+	return sec_part_enter_user_mode(spc);
+}
+
+TEE_Result sec_part_init_session(const TEE_UUID *uuid,
+				 struct tee_ta_session *sess)
+{
+	struct sec_part_ctx *spc = NULL;
+	TEE_Result res = TEE_SUCCESS;
+
+	if (memcmp(uuid, &stmm_uuid, sizeof(*uuid)))
+		return TEE_ERROR_ITEM_NOT_FOUND;
+
+	spc = sec_part_alloc_ctx(uuid);
+	if (!spc)
+		return TEE_ERROR_OUT_OF_MEMORY;
+
+	spc->is_initializing = true;
+
+	sess->ctx = &spc->uctx.ctx;
+	tee_ta_push_current_session(sess);
+	res = load_stmm(spc);
+	tee_ta_pop_current_session();
+	tee_mmu_set_ctx(NULL);
+	if (res)
+		goto err;
+
+	spc->is_initializing = false;
+	TAILQ_INSERT_TAIL(&tee_ctxes, &spc->uctx.ctx, link);
+	return TEE_SUCCESS;
+
+err:
+	sess->ctx = NULL;
+	spc->uctx.ctx.ops->destroy(&spc->uctx.ctx);
+
+	return res;
+}
+
+static TEE_Result stmm_enter_open_session(struct tee_ta_session *s,
+					  struct tee_ta_param *param __unused,
+					  TEE_ErrorOrigin *eo)
+{
+	struct sec_part_ctx *spc = to_sec_part_ctx(s->ctx);
+
+	if (spc->is_initializing) {
+		/* stmm is initialized in sec_part_init_session() */
+		*eo = TEE_ORIGIN_TEE;
+		return TEE_ERROR_BAD_STATE;
+	}
+
```

Comment:
We may want to check `param->types` that no parameters are passed, not super important but it might be a sign that the client isn't completely on top of things.